tests/functional/syntax/modules/test_initializers.py

```diff
@@ -0,0 +1,942 @@
+"""
+tests for the uses/initializes checker
+main properties to test:
+- state usage -- if a module uses state, it must `used` or `initialized`
+- conversely, if a module does not touch state, it should not be `used`
+- global initializer check: each used module is `initialized` exactly once
+"""
+
+import pytest
+
+from vyper.compiler import compile_code
+from vyper.exceptions import (
+    BorrowException,
+    ImmutableViolation,
+    InitializerException,
+    StructureException,
+    UndeclaredDefinition,
+)
+
+
+def test_initialize_uses(make_input_bundle):
+    lib1 = """
+counter: uint256
+
+@deploy
+def __init__():
+    pass
+    """
+    lib2 = """
+import lib1
+
+uses: lib1
+
+counter: uint256
+
+@deploy
+def __init__():
+    pass
+
+@internal
+def foo():
+    lib1.counter += 1
+    """
+    main = """
+import lib2
+import lib1
+
+initializes: lib2[lib1 := lib1]
+initializes: lib1
+
+@deploy
+def __init__():
+    lib1.__init__()
+    lib2.__init__()
+    """
+    input_bundle = make_input_bundle({"lib1.vy": lib1, "lib2.vy": lib2})
+
+    assert compile_code(main, input_bundle=input_bundle) is not None
+
+
+def test_initialize_uses_attribute(make_input_bundle):
+    lib1 = """
+counter: uint256
+
+@deploy
+def __init__():
+    pass
+    """
+    lib2 = """
+import lib1
+
+uses: lib1
+
+counter: uint256
+
+@deploy
+def __init__():
+    pass
+
+@internal
+def foo():
+    lib1.counter += 1
+    """
+    main = """
+import lib1
+import lib2
+
+initializes: lib2[lib1 := lib1]
+initializes: lib1
+
+@deploy
+def __init__():
+    lib2.__init__()
+    # demonstrate we can call lib1.__init__ through lib2.lib1
+    # (not sure this should be allowed, really.
+    lib2.lib1.__init__()
+    """
+    input_bundle = make_input_bundle({"lib1.vy": lib1, "lib2.vy": lib2})
+
+    assert compile_code(main, input_bundle=input_bundle) is not None
+
+
+def test_initializes_without_init_function(make_input_bundle):
+    lib1 = """
+counter: uint256
+    """
+    lib2 = """
+import lib1
+
+uses: lib1
+
+counter: uint256
+
+@internal
+def foo():
+    lib1.counter += 1
+    """
+    main = """
+import lib1
+import lib2
+
+initializes: lib2[lib1 := lib1]
+initializes: lib1
+
+@deploy
+def __init__():
+    pass
+    """
+    input_bundle = make_input_bundle({"lib1.vy": lib1, "lib2.vy": lib2})
+
+    assert compile_code(main, input_bundle=input_bundle) is not None
+
+
+def test_imported_as_different_names(make_input_bundle):
+    lib1 = """
+counter: uint256
+    """
+    lib2 = """
+import lib1 as m
+
+uses: m
+
+counter: uint256
+
+@internal
+def foo():
+    m.counter += 1
+    """
+    main = """
+import lib1
```

Comment:
Can this be imported as renamed as well? (at least in another test)

Reply:
hm, for the purposes of demonstration, why not

Reply:
cf. https://github.com/vyperlang/vyper/pull/3729/commits/353699cc6bfca72b567480c6ddac02e23da4733f